### PR TITLE
Document the v1.x guarantees

### LINKED ISSUE #2247: Document compatibility guarantees for v1
- Config files fields
- Flags
- Whats experimental & whats not.

### PR DESCRIPTION
**Pull request by gouthamve**
Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #2247

### CODE REVIEW

File: docs/configuration/v1-guarantees.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+---
+title: "v1.x Guarantees"
+linkTitle: "v1.x Guarantees"
+weight: 6 
+slug: v1guarantees
+---
+
+For the v1.0 release, we want to provide the following guarantees:
+
+# Flags, Config and minor version upgrades.
```
Comment (pracucci) [inline docs/configuration/v1-guarantees.md]:
```suggestion
# Flags, Config and minor version upgrades
```

File: docs/configuration/v1-guarantees.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+Upgrading cortex from one minor version to the next should "just work"; that being said, we don't want to bump the major version every time we remove a flag, so we will will keep deprecated flags around for 2 minor release.  There is a metric (`cortex_deprecated_flags_inuse_total`) you can alert on to find out if you're using a deprecated  flag.
+
+Similarly to flags, minor version upgrades using config files should "just work".  If we do need to change config, we will keep the old way working for two minor version.  There will be a metric you can alert on for this too.
+
+These gaurantees don't apply for [experimental features](#experimental-features).
```
Comment (pracucci) [inline docs/configuration/v1-guarantees.md]:
```suggestion
These guarantees don't apply for [experimental features](#experimental-features).
```

File: docs/configuration/v1-guarantees.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+# API Compatibility
+
+Cortex strives to be 100% API compatible with Prometheus (under `/api/prom/*`); any deviation from this is considered a bug, except:
+- Requiring the `__name__` label on queries.
+- Additional API endpoints for creating, removing and modifying alerts and recording rules.
+- Additional API around pushing metrics (under `/api/push`).
+- Additional API endpoints for management of Cortex itself, such as the ring.  These APIs are not part of the any compatibility guarantees.
+
+# Experimental features
```
Comment (khaines) [inline docs/configuration/v1-guarantees.md]: Should we make a pass on these to ensure all are prefixed with "experimental"? I realize that is a breaking change in itself, but I do like the convention that everything non-prod ready is prefixed with something consistent and obvious. 

Another option is to output a warning in logs on startup of experimental features, similar to what we do for deprecated flags. 

At the very least using the log output method, we could leave these flags alone for now and only force new experimental features to adhere to the convention.

File: docs/configuration/v1-guarantees.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+These gaurantees don't apply for [experimental features](#experimental-features).
+
+# Reading old data
+
+The Cortex maintainers commit to ensuring future version of Cortex can read data written by versions up to two years old. In practice we expect to be able to read more, but this is our guarantee.
+
+# API Compatibility
+
+Cortex strives to be 100% API compatible with Prometheus (under `/api/prom/*`); any deviation from this is considered a bug, except:
+- Requiring the `__name__` label on queries.
```
Comment (pracucci) [inline docs/configuration/v1-guarantees.md]: Add a new line before this, in order to the markdown list correctly compiled in the website

File: docs/configuration/v1-guarantees.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+- Additional API endpoints for creating, removing and modifying alerts and recording rules.
+- Additional API around pushing metrics (under `/api/push`).
+- Additional API endpoints for management of Cortex itself, such as the ring.  These APIs are not part of the any compatibility guarantees.
+
+# Experimental features
+
+Cortex is an actively developed project and we want to encourage the introduction of new features and capability.  As such, not everything in each release of Cortex is considered "production-ready". We don't provide any backwards compatibility guarantees on these and the config and flags might break.
+
+Examples of features considered experimental:
+- TSDB block storage.
```
Comment (pracucci) [inline docs/configuration/v1-guarantees.md]: Add a newline before this, please

File: docs/configuration/v1-guarantees.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+- Requiring the `__name__` label on queries.
+- Additional API endpoints for creating, removing and modifying alerts and recording rules.
+- Additional API around pushing metrics (under `/api/push`).
+- Additional API endpoints for management of Cortex itself, such as the ring.  These APIs are not part of the any compatibility guarantees.
+
+# Experimental features
+
+Cortex is an actively developed project and we want to encourage the introduction of new features and capability.  As such, not everything in each release of Cortex is considered "production-ready". We don't provide any backwards compatibility guarantees on these and the config and flags might break.
+
+Examples of features considered experimental:
```
Comment (pracucci) [inline docs/configuration/v1-guarantees.md]: I think the official documentation should have a clear idea about what is considered experimental. What if we remove "Examples of" and we replace this with:

```suggestion
Currently experimental features are:
```

File: docs/configuration/v1-guarantees.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+- TSDB block storage.
+- Ingester chunk WAL.
+- Cassandra storage engine.
+- Azure blob storage.
+- Zone awareness based replication.
+- Gossip based ring.
+- User subrings.
+- Ruler API (to PUT rules).
+- Memcached client DNS-based service discovery
```
Comment (pracucci) [inline docs/configuration/v1-guarantees.md]:
```suggestion
- Memcached client DNS-based service discovery.
```

File: docs/configuration/v1-guarantees.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+- Delete series APIs.
+- FIFO and Redis cache.
```
Comment (pracucci) [inline docs/configuration/v1-guarantees.md]:
```suggestion
- In-memory (FIFO) and Redis cache.
```

File: docs/configuration/v1-guarantees.md (hunk 3 above)
Comment (gouthamve) [reply docs/configuration/v1-guarantees.md]: See: https://github.com/cortexproject/cortex/pull/2361

File: docs/configuration/v1-guarantees.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+linkTitle: "v1.x Guarantees"
+weight: 6 
+slug: v1guarantees
+---
+
+For the v1.0 release, we want to provide the following guarantees:
+
+# Flags, Config and minor version upgrades
+
+Upgrading cortex from one minor version to the next should "just work"; that being said, we don't want to bump the major version every time we remove a flag, so we will will keep deprecated flags around for 2 minor release.  There is a metric (`cortex_deprecated_flags_inuse_total`) you can alert on to find out if you're using a deprecated  flag.
```
Comment (VineethReddy02) [inline docs/configuration/v1-guarantees.md]: @gouthamve _`so we will will`_ typo.

**Pull request merged**